File: src/docs_crawler/discovery.py
import json
import os
import re
import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from typing import Iterable, List, Optional, Set
from urllib.parse import quote, urlparse, urlunparse

from .http_client import fetch_bytes, fetch_text
from .models import EntryPointConfig
from .util import (
    join_url,
    normalize_url,
    split_language_from_path,
    url_has_allowed_extension,
)


ORDERED_LINK_RE = re.compile(r"\[[^\]]*\]\(([^)]+)\)|https?://[^\s)]+")
GITHUB_API_BASE_URL = "https://api.github.com"
GITHUB_RAW_BASE_URL = "https://raw.githubusercontent.com"


class DiscoveryError(RuntimeError):
    pass


@dataclass(frozen=True)
class DiscoveredDoc:
    url: str
    language_hint: Optional[str] = None
    relative_path: Optional[PurePosixPath] = None


def discover_documents(
    entrypoint: EntryPointConfig,
    *,
    allowed_extensions: Iterable[str],
    timeout_seconds: int,
    user_agent: str,
    default_language: str,
    cache_dir: Optional[os.PathLike] = None,
) -> List[DiscoveredDoc]:
    if entrypoint.type == "llms_txt":
        return _discover_llms_txt(
            entrypoint,
            allowed_extensions=allowed_extensions,
            timeout_seconds=timeout_seconds,
            user_agent=user_agent,
        )
    if entrypoint.type == "github_tree":
        return _discover_github_tree(
            entrypoint,
            allowed_extensions=allowed_extensions,
            timeout_seconds=timeout_seconds,
            user_agent=user_agent,
            default_language=default_language,
            cache_dir=cache_dir,
        )
    raise DiscoveryError(f"Unsupported entrypoint type: {entrypoint.type}")


def _discover_llms_txt(
    entrypoint: EntryPointConfig,
    *,
    allowed_extensions: Iterable[str],
    timeout_seconds: int,
    user_agent: str,
) -> List[DiscoveredDoc]:
    if not entrypoint.url:
        raise DiscoveryError("llms_txt entrypoint requires url")

    text = fetch_text(entrypoint.url, timeout_seconds=timeout_seconds, user_agent=user_agent)
    urls: Set[str] = set()
    ordered: List[DiscoveredDoc] = []
    base_language = entrypoint.language.strip() if entrypoint.language else None
    language_hint = base_language.lower() if base_language else None
    variant_languages = entrypoint.language_variants or None
    if variant_languages and not base_language:
        raise DiscoveryError("language_variants requires entrypoint.language")

    for line in text.splitlines():
        match = ORDERED_LINK_RE.search(line)
        if match is None:
            continue
        dest = match.group(1) if match.group(1) is not None else match.group(0)
        dest = dest.strip()
        if not dest or dest.startswith("#"):
            continue
        if match.group(1) is not None:
            dest = dest.split()[0].strip().strip("\"'")  # drop optional title
            if not dest or dest.startswith("#"):
                continue
            abs_url = join_url(entrypoint.url, dest)
        else:
            abs_url = normalize_url(dest)

        parsed = urlparse(abs_url)
        if parsed.scheme and parsed.scheme not in ("http", "https"):
            continue
        if not url_has_allowed_extension(abs_url, allowed_extensions):
            continue

        abs_url = normalize_url(abs_url)
        if variant_languages:
            if abs_url not in urls:
                urls.add(abs_url)
                ordered.append(DiscoveredDoc(url=abs_url, language_hint=language_hint))
            for variant in variant_languages:
                if not variant or not variant.strip():
                    continue
                variant_clean = variant.strip()
                if variant_clean.lower() == base_language.lower():
                    continue
                variant_url = _rewrite_language_segment(abs_url, base_language, variant_clean)
                if not variant_url:
                    continue
                variant_url = normalize_url(variant_url)
                if variant_url in urls:
                    continue
                urls.add(variant_url)
                ordered.append(DiscoveredDoc(url=variant_url, language_hint=variant_clean.lower()))
            continue

        if abs_url in urls:
            continue
        urls.add(abs_url)
        ordered.append(DiscoveredDoc(url=abs_url, language_hint=language_hint))

    return ordered


def _discover_github_tree(
    entrypoint: EntryPointConfig,
    *,
    allowed_extensions: Iterable[str],
    timeout_seconds: int,
    user_agent: str,
    default_language: str,
    cache_dir: Optional[os.PathLike],
) -> List[DiscoveredDoc]:
    if not entrypoint.repo or not entrypoint.path:
        raise DiscoveryError("github_tree entrypoint requires repo and path")

    repo = entrypoint.repo.strip().strip("/")
    ref = entrypoint.ref.strip() if entrypoint.ref else "main"
    base_path = entrypoint.path.strip().strip("/")
    api_base_url = (
        entrypoint.api_base_url.strip().rstrip("/") if entrypoint.api_base_url else GITHUB_API_BASE_URL
    )
    raw_base_url = (
        entrypoint.raw_base_url.strip().rstrip("/") if entrypoint.raw_base_url else GITHUB_RAW_BASE_URL
    )
    api_ref = quote(ref, safe="")
    api_url = f"{api_base_url}/repos/{repo}/git/trees/{api_ref}?recursive=1"

    try:
        payload = _fetch_json(api_url, timeout_seconds=timeout_seconds, user_agent=user_agent)
        if payload.get("truncated"):
            raise DiscoveryError(f"GitHub tree response truncated for {repo}@{ref}")

        tree = payload.get("tree")
        if not isinstance(tree, list):
            raise DiscoveryError("GitHub tree response missing tree list")
        tree_paths = [item.get("path") for item in tree if isinstance(item, dict) and item.get("type") == "blob"]
    except DiscoveryError as exc:
        try:
            tree_paths = _list_repo_paths_via_git(repo, ref, cache_dir=cache_dir)
        except DiscoveryError as git_exc:
            raise DiscoveryError(f"{exc}; git fallback failed: {git_exc}") from git_exc

    base_language = entrypoint.language.strip().lower() if entrypoint.language else default_language.strip().lower()
    allow_languages = {lang.strip().lower() for lang in entrypoint.languages or [] if lang.strip()}

    ordered: List[DiscoveredDoc] = []
    seen: Set[str] = set()
    for item_path in tree_paths:
        if not item_path or not isinstance(item_path, str):
            continue

        if base_path:
            if item_path == base_path:
                rel_path = PurePosixPath(PurePosixPath(base_path).name)
            elif item_path.startswith(f"{base_path}/"):
                rel_path = PurePosixPath(item_path[len(base_path) + 1 :])
            else:
                continue
        else:
            rel_path = PurePosixPath(item_path)

        raw_url = f"{raw_base_url}/{repo}/{ref}/{quote(item_path, safe='/')}"
        if not url_has_allowed_extension(raw_url, allowed_extensions):
            continue

        if allow_languages:
            language, _ = split_language_from_path(rel_path, base_language)
            if language.lower() not in allow_languages:
                continue

        if raw_url in seen:
            continue
        seen.add(raw_url)
        ordered.append(
            DiscoveredDoc(
                url=raw_url,
                language_hint=base_language,
                relative_path=rel_path,
            )
        )

    ordered.sort(key=lambda doc: doc.relative_path.as_posix() if doc.relative_path else doc.url)
    return ordered


def _list_repo_paths_via_git(
    repo: str,
    ref: str,
    *,
    cache_dir: Optional[os.PathLike] = None,
) -> List[str]:
    if shutil.which("git") is None:
        raise DiscoveryError("git is not available for github_tree fallback")

    remote_url = f"https://github.com/{repo}.git"
    env = dict(os.environ)
    env.setdefault("GIT_TERMINAL_PROMPT", "0")
    env.setdefault("GIT_HTTP_VERSION", "HTTP/1.1")
    token = os.environ.get("GITHUB_TOKEN") or os.environ.get("GH_TOKEN")
    fetch_cmd = ["git"]
    if token:
        fetch_cmd.extend(["-c", f"http.extraHeader=Authorization: Bearer {token}"])
    fetch_cmd.extend(["fetch", "--depth", "1", "origin", ref])

    if cache_dir is None:
        with tempfile.TemporaryDirectory(prefix="docs-crawler-git-") as tmpdir:
            _run_git(["git", "init", "-q"], cwd=tmpdir, env=env)
            _run_git(["git", "remote", "add", "origin", remote_url], cwd=tmpdir, env=env)
            _run_git(fetch_cmd, cwd=tmpdir, env=env)
            result = _run_git(
                ["git", "ls-tree", "-r", "--name-only", "FETCH_HEAD"],
                cwd=tmpdir,
                env=env,
                capture_output=True,
            )
        return [line.strip() for line in result.stdout.splitlines() if line.strip()]

    repo_dir = _prepare_git_cache_dir(cache_dir, repo, remote_url, env)
    _run_git(fetch_cmd, cwd=str(repo_dir), env=env)
    result = _run_git(
        ["git", "ls-tree", "-r", "--name-only", "FETCH_HEAD"],
        cwd=str(repo_dir),
        env=env,
        capture_output=True,
    )
    return [line.strip() for line in result.stdout.splitlines() if line.strip()]


def _prepare_git_cache_dir(
    cache_dir: os.PathLike,
    repo: str,
    remote_url: str,
    env: dict,
) -> Path:
    cache_root = Path(cache_dir).expanduser()
    parts = [part for part in repo.split("/") if part not in ("", ".", "..")]
    repo_dir = cache_root.joinpath(*parts)
    repo_dir.mkdir(parents=True, exist_ok=True)

    try:
        _run_git(["git", "rev-parse", "--is-inside-work-tree"], cwd=str(repo_dir), env=env)
    except DiscoveryError:
        _run_git(["git", "init", "-q"], cwd=str(repo_dir), env=env)

    try:
        result = _run_git(
            ["git", "remote", "get-url", "origin"],
            cwd=str(repo_dir),
            env=env,
            capture_output=True,
        )
        current_remote = result.stdout.strip()
        if current_remote and current_remote != remote_url:
            _run_git(["git", "remote", "set-url", "origin", remote_url], cwd=str(repo_dir), env=env)
    except DiscoveryError:
        _run_git(["git", "remote", "add", "origin", remote_url], cwd=str(repo_dir), env=env)

    return repo_dir


def _run_git(
    args: List[str],
    *,
    cwd: str,
    env: dict,
    capture_output: bool = False,
) -> subprocess.CompletedProcess:
    try:
        return subprocess.run(
            args,
            cwd=cwd,
            env=env,
            check=True,
            text=True,
            stdout=subprocess.PIPE if capture_output else subprocess.DEVNULL,
            stderr=subprocess.PIPE,
        )
    except FileNotFoundError as exc:
        raise DiscoveryError("git is not installed") from exc
    except subprocess.CalledProcessError as exc:
        detail = exc.stderr.strip() or "unknown git error"
        raise DiscoveryError(f"git command failed: {detail}") from exc


def _fetch_json(
    url: str,
    *,
    timeout_seconds: int,
    user_agent: str,
) -> dict:
    token = os.environ.get("GITHUB_TOKEN") or os.environ.get("GH_TOKEN")
    extra_headers = {"Authorization": f"Bearer {token}"} if token else None
    result = fetch_bytes(
        url,
        timeout_seconds=timeout_seconds,
        user_agent=user_agent,
        accept="application/vnd.github+json",
        extra_headers=extra_headers,
    )
    if result.status != 200 or result.body is None:
        raise DiscoveryError(f"Failed to fetch {url}: status={result.status} error={result.error}")
    try:
        return json.loads(result.body.decode("utf-8"))
    except json.JSONDecodeError as exc:  # pragma: no cover
        raise DiscoveryError(f"Invalid JSON from {url}") from exc


def _rewrite_language_segment(url: str, base_language: str, target_language: str) -> Optional[str]:
    parsed = urlparse(url)
    parts = parsed.path.split("/")
    replaced = False
    for idx, part in enumerate(parts):
        if part.lower() == base_language.lower():
            parts[idx] = target_language
            replaced = True
            break
    if not replaced:
        return None
    return urlunparse(parsed._replace(path="/".join(parts)))

File: crates/docs-crawler/src/discovery.rs
use crate::http_client::HttpClient;
use crate::models::EntryPointConfig;
use crate::util::{join_url, normalize_url, split_language_from_path, url_has_allowed_extension};
use once_cell::sync::Lazy;
use percent_encoding::AsciiSet;
use percent_encoding::CONTROLS;
use regex::Regex;
use serde::Deserialize;
use std::collections::{HashMap, HashSet};
use std::env;
use std::error::Error;
use std::fmt::{self, Display};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use tempfile::TempDir;
use url::Url;

static ORDERED_LINK_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"\[[^\]]*\]\(([^)]+)\)|https?://[^\s)]+").expect("valid llms.txt regex")
});

const GITHUB_API_BASE_URL: &str = "https://api.github.com";
const GITHUB_RAW_BASE_URL: &str = "https://raw.githubusercontent.com";

#[derive(Debug)]
pub struct DiscoveryError(String);

impl Display for DiscoveryError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Error for DiscoveryError {}

#[derive(Debug, Clone)]
pub struct DiscoveredDoc {
    pub url: String,
    pub language_hint: Option<String>,
    pub relative_path: Option<PathBuf>,
}

pub fn discover_documents(
    entrypoint: &EntryPointConfig,
    allowed_extensions: &[String],
    http: &HttpClient,
    default_language: &str,
    cache_dir: Option<&Path>,
) -> Result<Vec<DiscoveredDoc>, DiscoveryError> {
    match entrypoint.kind.as_str() {
        "llms_txt" => discover_llms_txt(entrypoint, allowed_extensions, http),
        "github_tree" => discover_github_tree(
            entrypoint,
            allowed_extensions,
            http,
            default_language,
            cache_dir,
        ),
        other => Err(DiscoveryError(format!(
            "Unsupported entrypoint type: {other}"
        ))),
    }
}

fn discover_llms_txt(
    entrypoint: &EntryPointConfig,
    allowed_extensions: &[String],
    http: &HttpClient,
) -> Result<Vec<DiscoveredDoc>, DiscoveryError> {
    let url = entrypoint
        .url
        .as_ref()
        .ok_or_else(|| DiscoveryError("llms_txt entrypoint requires url".to_string()))?;
    let text = http
        .fetch_text(url)
        .map_err(|err| DiscoveryError(err.to_string()))?;

    let base_language = entrypoint
        .language
        .as_ref()
        .map(|lang| lang.trim().to_string());
    let language_hint = base_language.as_ref().map(|lang| lang.to_lowercase());
    let variant_languages = entrypoint.language_variants.as_ref();
    if variant_languages.is_some() && base_language.is_none() {
        return Err(DiscoveryError(
            "language_variants requires entrypoint.language".to_string(),
        ));
    }

    let mut urls = HashSet::new();
    let mut ordered = Vec::new();

    for line in text.lines() {
        let Some(captures) = ORDERED_LINK_RE.captures(line) else {
            continue;
        };
        let dest = if let Some(group) = captures.get(1) {
            group.as_str()
        } else {
            captures.get(0).map(|m| m.as_str()).unwrap_or("")
        };
        let mut dest = dest.trim();
        if dest.is_empty() || dest.starts_with('#') {
            continue;
        }
        let abs_url = if captures.get(1).is_some() {
            dest = dest.split_whitespace().next().unwrap_or("");
            dest = dest.trim_matches(&['"', '\''][..]);
            if dest.is_empty() || dest.starts_with('#') {
                continue;
            }
            join_url(url, dest).map_err(DiscoveryError)?
        } else {
            normalize_url(dest)
        };

        let parsed = Url::parse(&abs_url);
        if let Ok(parsed) = parsed {
            if parsed.scheme() != "http" && parsed.scheme() != "https" {
                continue;
            }
        }

        if !url_has_allowed_extension(&abs_url, allowed_extensions) {
            continue;
        }

        let abs_url = normalize_url(&abs_url);
        if let Some(variants) = variant_languages {
            if urls.insert(abs_url.clone()) {
                ordered.push(DiscoveredDoc {
                    url: abs_url.clone(),
                    language_hint: language_hint.clone(),
                    relative_path: None,
                });
            }
            let base_language = base_language.as_ref().unwrap();
            for variant in variants {
                let variant_clean = variant.trim();
                if variant_clean.is_empty() {
                    continue;
                }
                if variant_clean.eq_ignore_ascii_case(base_language) {
                    continue;
                }
                let Some(variant_url) =
                    rewrite_language_segment(&abs_url, base_language, variant_clean)
                else {
                    continue;
                };
                let variant_url = normalize_url(&variant_url);
                if urls.insert(variant_url.clone()) {
                    ordered.push(DiscoveredDoc {
                        url: variant_url,
                        language_hint: Some(variant_clean.to_lowercase()),
                        relative_path: None,
                    });
                }
            }
            continue;
        }

        if urls.insert(abs_url.clone()) {
            ordered.push(DiscoveredDoc {
                url: abs_url,
                language_hint: language_hint.clone(),
                relative_path: None,
            });
        }
    }

    Ok(ordered)
}

fn discover_github_tree(
    entrypoint: &EntryPointConfig,
    allowed_extensions: &[String],
    http: &HttpClient,
    default_language: &str,
    cache_dir: Option<&Path>,
) -> Result<Vec<DiscoveredDoc>, DiscoveryError> {
    let repo = entrypoint
        .repo
        .as_ref()
        .ok_or_else(|| DiscoveryError("github_tree entrypoint requires repo".to_string()))?
        .trim()
        .trim_matches('/');
    let r#ref = entrypoint
        .r#ref
        .as_deref()
        .unwrap_or("main")
        .trim()
        .to_string();
    let base_path = entrypoint
        .path
        .as_ref()
        .ok_or_else(|| DiscoveryError("github_tree entrypoint requires path".to_string()))?
        .trim()
        .trim_matches('/')
        .to_string();
    let api_base_url = entrypoint
        .api_base_url
        .as_deref()
        .unwrap_or(GITHUB_API_BASE_URL)
        .trim()
        .trim_end_matches('/');
    let raw_base_url = entrypoint
        .raw_base_url
        .as_deref()
        .unwrap_or(GITHUB_RAW_BASE_URL)
        .trim()
        .trim_end_matches('/');

    let api_ref = percent_encode_component(&r#ref);
    let api_url = format!("{api_base_url}/repos/{repo}/git/trees/{api_ref}?recursive=1");

    let tree_paths = match fetch_github_tree(&api_url, http) {
        Ok(paths) => paths,
        Err(err) => list_repo_paths_via_git(repo, &r#ref, cache_dir)
            .map_err(|git_err| DiscoveryError(format!("{err}; git fallback failed: {git_err}")))?,
    };

    let base_language = entrypoint
        .language
        .as_deref()
        .unwrap_or(default_language)
        .trim()
        .to_lowercase();
    let allow_languages: HashSet<String> = entrypoint
        .languages
        .clone()
        .unwrap_or_default()
        .into_iter()
        .filter_map(|lang| {
            let trimmed = lang.trim().to_lowercase();
            if trimmed.is_empty() {
                None
            } else {
                Some(trimmed)
            }
        })
        .collect();

    let mut ordered = Vec::new();
    let mut seen = HashSet::new();

    for item_path in tree_paths {
        if item_path.is_empty() {
            continue;
        }

        let rel_path = if base_path.is_empty() {
            PathBuf::from(&item_path)
        } else if item_path == base_path {
            match Path::new(&base_path).file_name() {
                Some(name) => PathBuf::from(name),
                None => PathBuf::from(&base_path),
            }
        } else if item_path.starts_with(&format!("{base_path}/")) {
            PathBuf::from(&item_path[base_path.len() + 1..])
        } else {
            continue;
        };

        let raw_url =
            build_raw_url(raw_base_url, repo, &r#ref, &item_path).map_err(DiscoveryError)?;

        if !url_has_allowed_extension(&raw_url, allowed_extensions) {
            continue;
        }

        if !allow_languages.is_empty() {
            let (language, _) = split_language_from_path(&rel_path, &base_language, None);
            if !allow_languages.contains(&language) {
                continue;
            }
        }

        if seen.insert(raw_url.clone()) {
            ordered.push(DiscoveredDoc {
                url: raw_url,
                language_hint: Some(base_language.clone()),
                relative_path: Some(rel_path),
            });
        }
    }

    ordered.sort_by(|a, b| {
        let a_key = a
            .relative_path
            .as_ref()
            .map(|path| path.to_string_lossy().to_string())
            .unwrap_or_else(|| a.url.clone());
        let b_key = b
            .relative_path
            .as_ref()
            .map(|path| path.to_string_lossy().to_string())
            .unwrap_or_else(|| b.url.clone());
        a_key.cmp(&b_key)
    });

    Ok(ordered)
}

fn percent_encode_component(value: &str) -> String {
    const ENCODE_SET: &AsciiSet = &CONTROLS
        .add(b' ')
        .add(b'"')
        .add(b'#')
        .add(b'<')
        .add(b'>')
        .add(b'?')
        .add(b'`')
        .add(b'{')
        .add(b'}')
        .add(b'%')
        .add(b'+');
    percent_encoding::utf8_percent_encode(value, ENCODE_SET).to_string()
}

fn build_raw_url(
    raw_base_url: &str,
    repo: &str,
    r#ref: &str,
    item_path: &str,
) -> Result<String, String> {
    let mut base = Url::parse(&format!("{raw_base_url}/{repo}/{ref}/"))
        .map_err(|err| format!("Invalid raw base url: {err}"))?;
    {
        let mut segments = base
            .path_segments_mut()
            .map_err(|_| "Cannot set raw URL path".to_string())?;
        segments.pop_if_empty();
        for segment in item_path.split('/') {
            if segment.is_empty() {
                continue;
            }
            segments.push(segment);
        }
    }
    Ok(base.to_string())
}

#[derive(Debug, Deserialize)]
struct GitTreeResponse {
    truncated: Option<bool>,
    tree: Option<Vec<GitTreeItem>>,
}

#[derive(Debug, Deserialize)]
struct GitTreeItem {
    path: Option<String>,
    #[serde(rename = "type")]
    kind: Option<String>,
}

fn fetch_github_tree(api_url: &str, http: &HttpClient) -> Result<Vec<String>, DiscoveryError> {
    let mut headers = HashMap::new();
    if let Ok(token) = env::var("GITHUB_TOKEN").or_else(|_| env::var("GH_TOKEN")) {
        headers.insert("Authorization".to_string(), format!("Bearer {token}"));
    }

    let result = http.fetch_bytes(
        api_url,
        Some("application/vnd.github+json"),
        None,
        None,
        if headers.is_empty() {
            None
        } else {
            Some(&headers)
        },
    );
    if result.status != 200 || result.body.is_none() {
        return Err(DiscoveryError(format!(
            "Failed to fetch {}: status={} error={}",
            api_url,
            result.status,
            result.error.unwrap_or_default()
        )));
    }

    let response: GitTreeResponse = serde_json::from_slice(result.body.as_ref().unwrap())
        .map_err(|_| DiscoveryError(format!("Invalid JSON from {api_url}")))?;
    if response.truncated.unwrap_or(false) {
        return Err(DiscoveryError(format!(
            "GitHub tree response truncated for {api_url}"
        )));
    }
    let tree = response
        .tree
        .ok_or_else(|| DiscoveryError("GitHub tree response missing tree list".to_string()))?;

    let mut paths = Vec::new();
    for item in tree {
        if item.kind.as_deref() != Some("blob") {
            continue;
        }
        if let Some(path) = item.path {
            paths.push(path);
        }
    }
    Ok(paths)
}

fn list_repo_paths_via_git(
    repo: &str,
    r#ref: &str,
    cache_dir: Option<&Path>,
) -> Result<Vec<String>, DiscoveryError> {
    let remote_url = format!("https://github.com/{repo}.git");
    let mut envs = HashMap::new();
    envs.insert("GIT_TERMINAL_PROMPT".to_string(), "0".to_string());
    envs.insert("GIT_HTTP_VERSION".to_string(), "HTTP/1.1".to_string());

    let token = env::var("GITHUB_TOKEN")
        .or_else(|_| env::var("GH_TOKEN"))
        .ok();
    let mut fetch_args = Vec::new();
    if let Some(token) = token.as_deref() {
        fetch_args.push("-c".to_string());
        fetch_args.push(format!("http.extraHeader=Authorization: Bearer {token}"));
    }
    fetch_args.push("fetch".to_string());
    fetch_args.push("--depth".to_string());
    fetch_args.push("1".to_string());
    fetch_args.push("origin".to_string());
    fetch_args.push(r#ref.to_string());

    let output = if let Some(cache_dir) = cache_dir {
        let repo_dir = prepare_git_cache_dir(cache_dir, repo, &remote_url, &envs)?;
        run_git(&fetch_args, &repo_dir, &envs, false)?;
        run_git(
            &[
                "ls-tree".to_string(),
                "-r".to_string(),
                "--name-only".to_string(),
                "FETCH_HEAD".to_string(),
            ],
            &repo_dir,
            &envs,
            true,
        )?
    } else {
        let temp_root = TempDir::new()
            .map_err(|err| DiscoveryError(format!("Failed to create git temp dir: {err}")))?;
        let repo_dir = temp_root.path().join("repo");
        let mut clone_args = Vec::new();
        if let Some(token) = token.as_deref() {
            clone_args.push("-c".to_string());
            clone_args.push(format!("http.extraHeader=Authorization: Bearer {token}"));
        }
        clone_args.push("clone".to_string());
        clone_args.push("--depth".to_string());
        clone_args.push("1".to_string());
        clone_args.push("--branch".to_string());
        clone_args.push(r#ref.to_string());
        clone_args.push(remote_url.clone());
        clone_args.push(repo_dir.to_string_lossy().to_string());

        run_git(&clone_args, temp_root.path(), &envs, false)?;
        run_git(
            &[
                "ls-tree".to_string(),
                "-r".to_string(),
                "--name-only".to_string(),
                "HEAD".to_string(),
            ],
            &repo_dir,
            &envs,
            true,
        )?
    };

    Ok(output
        .lines()
        .map(|line| line.trim().to_string())
        .filter(|line| !line.is_empty())
        .collect())
}

fn prepare_git_cache_dir(
    cache_dir: &Path,
    repo: &str,
    remote_url: &str,
    envs: &HashMap<String, String>,
) -> Result<PathBuf, DiscoveryError> {
    let mut repo_dir = cache_dir.to_path_buf();
    for part in repo.split('/') {
        if part.is_empty() || part == "." || part == ".." {
            continue;
        }
        repo_dir.push(part);
    }
    fs::create_dir_all(&repo_dir)
        .map_err(|err| DiscoveryError(format!("Failed to create cache dir: {err}")))?;

    if run_git(
        &["rev-parse".to_string(), "--is-inside-work-tree".to_string()],
        &repo_dir,
        envs,
        false,
    )
    .is_err()
    {
        run_git(
            &["init".to_string(), "-q".to_string()],
            &repo_dir,
            envs,
            false,
        )?;
    }

    match run_git(
        &[
            "remote".to_string(),
            "get-url".to_string(),
            "origin".to_string(),
        ],
        &repo_dir,
        envs,
        true,
    ) {
        Ok(output) => {
            let current = output.trim();
            if !current.is_empty() && current != remote_url {
                run_git(
                    &[
                        "remote".to_string(),
                        "set-url".to_string(),
                        "origin".to_string(),
                        remote_url.to_string(),
                    ],
                    &repo_dir,
                    envs,
                    false,
                )?;
            }
        }
        Err(_) => {
            run_git(
                &[
                    "remote".to_string(),
                    "add".to_string(),
                    "origin".to_string(),
                    remote_url.to_string(),
                ],
                &repo_dir,
                envs,
                false,
            )?;
        }
    }

    Ok(repo_dir)
}

fn run_git(
    args: &[String],
    cwd: &Path,
    envs: &HashMap<String, String>,
    capture_output: bool,
) -> Result<String, DiscoveryError> {
    let mut command = Command::new("git");
    command.args(args).current_dir(cwd);
    for (key, value) in envs {
        command.env(key, value);
    }
    if capture_output {
        command.stdout(Stdio::piped()).stderr(Stdio::piped());
    } else {
        command.stdout(Stdio::null()).stderr(Stdio::piped());
    }

    let output = command
        .output()
        .map_err(|err| DiscoveryError(format!("git command failed to start: {err}")))?;
    if !output.status.success() {
        let detail = String::from_utf8_lossy(&output.stderr).trim().to_string();
        let detail = if detail.is_empty() {
            "unknown git error".to_string()
        } else {
            detail
        };
        return Err(DiscoveryError(format!("git command failed: {detail}")));
    }

    if capture_output {
        Ok(String::from_utf8_lossy(&output.stdout).to_string())
    } else {
        Ok(String::new())
    }
}

fn rewrite_language_segment(
    url: &str,
    base_language: &str,
    target_language: &str,
) -> Option<String> {
    let mut parsed = Url::parse(url).ok()?;
    let mut segments: Vec<String> = parsed
        .path_segments()
        .map(|segments| segments.map(|seg| seg.to_string()).collect())
        .unwrap_or_default();
    let mut replaced = false;
    for segment in segments.iter_mut() {
        if segment.eq_ignore_ascii_case(base_language) {
            *segment = target_language.to_string();
            replaced = true;
            break;
        }
    }
    if !replaced {
        return None;
    }
    let new_path = format!("/{}", segments.join("/"));
    parsed.set_path(&new_path);
    Some(parsed.to_string())
}
